src/main/flight/gps_rescue.c

```diff
@@ -0,0 +1,373 @@
+                rescueState.phase = RESCUE_CROSSTRACK;
+            }
+
+            rescueState.intent.targetGroundspeed = 500;
+            rescueState.intent.targetAltitude = MAX(gpsRescue()->initialAltitude * 100, rescueState.sensor.maxAltitude + 1500);
+            rescueState.intent.crosstrack = true;
+            rescueState.intent.minAngle = 10;
+            rescueState.intent.maxAngle = 15;
+            break;
+        case RESCUE_CROSSTRACK:
+            if (rescueState.sensor.distanceToHome < gpsRescue()->descentDistance) {
+                rescueState.phase = RESCUE_LANDING_APPROACH;
+            }
+
+            // We can assume at this point that we are at or above our RTH height, so we need to try and point to home and tilt while maintaining alt
+            // Is our altitude way off?  We should probably kick back to phase RESCUE_ATTAIN_ALT
+            rescueState.intent.targetGroundspeed = gpsRescue()->rescueGroundspeed;
+            rescueState.intent.targetAltitude = MAX(gpsRescue()->initialAltitude * 100, rescueState.sensor.maxAltitude + 1500);
+            rescueState.intent.crosstrack = true;
+            rescueState.intent.minAngle = 15;
+            rescueState.intent.maxAngle = gpsRescue()->angle;
+            break;
+        case RESCUE_LANDING_APPROACH:
+            // We are getting close to home in the XY plane, get Z where it needs to be to move to landing phase
+            if (rescueState.sensor.distanceToHome < 10 && rescueState.sensor.currentAltitude <= 1000) {
+                rescueState.phase = RESCUE_LANDING;
+            }
+
+            // Only allow new altitude and new speed to be equal or lower than the current values (to prevent parabolic movement on overshoot)
+            int32_t newAlt = gpsRescue()->initialAltitude * 100  * rescueState.sensor.distanceToHome / gpsRescue()->descentDistance;
+            int32_t newSpeed = gpsRescue()->rescueGroundspeed * rescueState.sensor.distanceToHome / gpsRescue()->descentDistance;
+
+            rescueState.intent.targetAltitude = constrain(newAlt, 100, rescueState.intent.targetAltitude);
+            rescueState.intent.targetGroundspeed = constrain(newSpeed, 100, rescueState.intent.targetGroundspeed);
+            rescueState.intent.crosstrack = true;
+            rescueState.intent.minAngle = 10;
+            rescueState.intent.maxAngle = 20;
+            break;
+        case RESCUE_LANDING:
+            // We have reached the XYZ envelope to be considered at "home".  We need to land gently and check our accelerometer for abnormal data.
+            // At this point, do not let the target altitude go up anymore, so if we overshoot, we dont' move in a parabolic trajectory
+
+            // If we are over 120% of average magnitude, just disarm since we're pretty much home
+            if (rescueState.sensor.accMagnitude > rescueState.sensor.accMagnitudeAvg * 1.5) {
```

Comment:
This is not noise safe, but rescue landing phase cannot be entered unless it is very close to home (XYZ envelope), so the idea was to make it extremely sensitive to anything happening to the acceleration and disarm, because even if it disarmed the moment it got to landing phase, wouldn't be a long fall :)